### PR TITLE
SAC-7 Delete User DELETE

### PR DESCRIPTION
**Pull request by in-mai-space**
# Description

[Link to Ticket](https://generatesac.atlassian.net/jira/software/projects/SAC/boards/1?selectedIssue=SAC-7)

- Implements DELETE for user

# How Has This Been Tested?

- Successful delete (200)
- User Does Not Exist (404)
- Invalid string ID (400)
- Negative ID (400)
- Database Disconnected (500)

# Checklist

- [x] I have performed a self-review of my code
- [x] I have reached out to another developer to review my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: backend/src/services/user.go [hunk 1]
```diff
@@ -19,3 +23,15 @@ type UserService struct {
 func (u *UserService) GetAllUsers() ([]models.User, error) {
 	return transactions.GetAllUsers(u.DB)
 }
+
+// Delete user with a specific id
+func (u *UserService) DeleteUser(id string) (error) {
+	idAsInt, err := strconv.Atoi(id)
```
Comment (DOOduneye) [inline backend/src/services/user.go:29]: Definitely something to be extracted into some validation function with how commonly this will be done

**Review by DOOduneye — CHANGES_REQUESTED** [review]
Looks great, just some small nitpicks

File: backend/src/controllers/user.go [hunk 2]
```diff
@@ -33,3 +33,23 @@ func (u *UserController) GetAllUsers(c *fiber.Ctx) error {
...
+// @Failure     500   {string}     string "failed to get all users"
+// @Router		/api/v1/users/:id  [delete]
+
+func (u * UserController) DeleteUser(c *fiber.Ctx) error {
+	userID := c.Params("id")
+	err := u.userService.DeleteUser(userID)
+	if err != nil {
+		return err
+	}
+	return nil
```
Comment (garrettladley) [inline backend/src/controllers/user.go:54]:
```suggestion
	return fiber.SendStatus(fiber.StatusNoContent)
```

File: backend/src/services/user.go (hunk 1 above)
Comment (garrettladley) [reply backend/src/services/user.go:29]: Let's leverage the `utilities.ValidateID` here

File: backend/src/transactions/user.go [hunk 3]
```diff
@@ -16,3 +18,18 @@ func GetAllUsers(db *gorm.DB) ([]models.User, error) {
 
 	return users, nil
 }
+
+func DeleteUser(db *gorm.DB, id uint) error {
+	var deletedUser models.User
+
+	result := db.Where("id = ?", id).Delete(&deletedUser)
+	if result.RowsAffected == 0 {
+		err := db.Where("id = ?", id).First(&deletedUser)
```
Comment (garrettladley) [inline backend/src/transactions/user.go:27]: can we get away without this second query?

File: backend/src/transactions/user.go [hunk 4]
```diff
@@ -16,3 +18,18 @@ func GetAllUsers(db *gorm.DB) ([]models.User, error) {
...
+		if errors.Is(err.Error, gorm.ErrRecordNotFound) {
+			return fiber.NewError(fiber.StatusNotFound, "user not found")
+		} else {
+			return fiber.NewError(fiber.StatusInternalServerError, "not connected to database")
```
Comment (garrettladley) [inline backend/src/transactions/user.go:31]: is this the only reason?

File: backend/tests/api/user_test.go [hunk 5]
```diff
@@ -45,3 +47,70 @@ func TestGetAllUsersWorks(t *testing.T) {
 		},
 	)
 }
+
+func TestDeleteUserWorks(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (garrettladley) [reply backend/tests/api/user_test.go:51]: yeah i will handle

**Review by garrettladley — CHANGES_REQUESTED** [review]
nits but great work team 👏

**Pull request merged**